README.md
# fltk-grid

A grid widget for fltk-rs.

## Usage
```toml
[dependencies]
fltk = "1.4.15"
fltk-grid = "0.4"
```

Basically, the crate contains a single type Grid which has 4 main non-constructor methods:
- set_layout(): specifies the number of rows and columns of the grid.
- set_widget(): specifies the widget to be placed, along with in which cell (row, column). The values can be a range (0..1).
- set_widget_ext(): adds to set_widget the row span and column span, in addition to the Grid Alignment.
- resize(): determines how the grid is resized.
- debug(): shows the cell outline and their numbering, useful for prototyping. 


```rust
use fltk::{prelude::*, *};
use fltk_grid::Grid;

fn main() {
    let a = app::App::default().with_scheme(app::Scheme::Gtk);
    let mut win = window::Window::default().with_size(500, 300);
    let mut grid = Grid::default_fill();
    grid.show_grid(false); // set to true to show cell outlines and numbers
    grid.set_layout(5, 5); // 5 rows, 5 columns
    grid.set_widget(&mut button::Button::default(), 0, 1); // widget, row, col
    grid.set_widget(&mut button::Button::default(), 2..3, 1..4); // widget, row range, col range
    // or
    // grid.set_widget_ext(&mut button::Button::default(), 2, 1, 1, 3, GridAlign::FILL); // widget, row, col, row_span, col_span
    grid.end();
    win.end();
    win.show();
    a.run().unwrap();
}
```

## Example
Run `cargo run --example form`

- [Form example](https://github.com/fltk-rs/fltk-grid/blob/main/examples/form.rs)

![image](https://user-images.githubusercontent.com/37966791/160347418-b8b54408-3dc9-4fc4-93e8-fb6c1c0282e9.png)

Setting Grid::debug(true):

![image](https://user-images.githubusercontent.com/37966791/160346084-f3b0dad7-bd14-41da-99a0-768f7327ab2c.png)


CHANGELOG.md
# [0.4] - Unreleased
- Use FLTK's new grid implementation.
- Change debug() to show_grid().
- Change insert and insert_ext to set_widget and set_widget_ext.
- Add set_layout_ext which takes a margin and a gap.
- Require the call to Grid::end().
- set_widget and set_widget_ext no longer add the passed widgets as children automatically.


examples/ex.rs
use fltk::{prelude::*, *};
use fltk_grid::Grid;

struct Form {
    grid: Grid,
    name: input::Input,
    age: input::IntInput,
    occupation: input::Input,
    btn: button::Button,
}

impl Form {
    pub fn default() -> Self {
        let mut grid = Grid::default_fill();
        grid.set_layout(10, 5); // construct a new grid
        let name = input::Input::default();
        let age = input::IntInput::default();
        let occupation = input::Input::default();
        let btn = button::Button::default().with_label("Submit");
        let mut g = Self {
            grid,
            name,
            age,
            occupation,
            btn,
        };
        g.fill();
        g
    }

    fn fill(&mut self) {
        let grid = &mut self.grid;
        grid.debug(true); // set to true to see cell outlines
        grid.insert_ext(
            // insert widgets
            &mut frame::Frame::default().with_label("Employee Form"),
            0,
            1,
            3,
            1,
        );
        grid.insert(&mut frame::Frame::default().with_label("Name"), 2, 1);
        grid.insert(&mut self.name, 2, 3);
        grid.insert(&mut frame::Frame::default().with_label("Age"), 4, 1);
        grid.insert(&mut self.age, 4, 3);
        grid.insert(&mut frame::Frame::default().with_label("Occupation"), 6, 1);
        grid.insert(&mut self.occupation, 6, 3);
        grid.insert(&mut self.btn, 8, 2);
    }

    fn register_default_callback(&mut self) {
        self.btn.set_callback({
            let name = self.name.clone();
            let age = self.age.clone();
            let occupation = self.occupation.clone();
            move |_| {
                println!("Name: {}", name.value());
                println!("Age: {}", age.value());
                println!("Occupation: {}", occupation.value());
            }
        });
    }

    pub fn resize(&mut self, x: i32, y: i32, w: i32, h: i32) {
        self.grid.resize(x, y, w, h); // determine how it's resized
    }
}

fn main() {
    let a = app::App::default();
    let mut win = window::Window::default().with_size(500, 400);
    let mut form = Form::default();
    form.register_default_callback();
    win.end();
    win.make_resizable(true);
    win.show();

    win.resize_callback(move |_, _, _, w, h| form.resize(0, 0, w, h));

    a.run().unwrap();
}


examples/simple.rs
use fltk::{prelude::*, *};
use fltk_grid::Grid;

fn main() {
    let a = app::App::default().with_scheme(app::Scheme::Gtk);
    let mut win = window::Window::default().with_size(500, 300);
    let mut grid = Grid::default_fill();
    grid.show_grid(false); // set to true to show cell outlines and numbers
    grid.set_layout(5, 5); // 5 rows, 5 columns
    grid.set_widget(&mut button::Button::default(), 0, 1); // widget, row, col
    grid.set_widget(&mut button::Button::default(), 2..3, 1..4); // widget, row range, col range
                                                             // or
                                                            //  grid.set_widget_ext(&mut button::Button::default(), 2, 1, 1, 3, GridAlign::FILL); // widget, row, col, row_span, col_span
    grid.end();
    win.end();
    win.show();
    a.run().unwrap();
    dbg!(grid.children());
}


examples/grid_in_grid.rs
#![allow(dead_code)]

use fltk::{prelude::*, *};
use fltk_grid::Grid;

struct Panel {
    grid: Grid,
    label: frame::Frame,
    cb1: button::CheckButton,
    cb2: button::CheckButton,
    btn: button::Button,
}
impl Panel {
    pub fn default() -> Self {
        let mut grid = Grid::default();
        grid.show_grid(true);
        grid.set_layout(6, 1);
        let mut label = frame::Frame::default().with_label("ARTERY:");

        let mut cb1 = button::CheckButton::default().with_label("Normal");
        let mut cb2 = button::CheckButton::default().with_label("Normal");

        let cbvec = vec![cb1.clone(), cb2.clone()];

        let mut btn = button::Button::default().with_label("Submit");
        grid.set_widget(&mut label, 0, 0);
        grid.set_widget(&mut cb1, 1, 0);
        grid.set_widget(&mut cb2, 2, 0);
        grid.set_widget(&mut btn, 5, 0);
        btn.set_callback(move |_btn| {
            for cb in cbvec.clone() {
                println!(
                    "CB status: {}",
                    if cb.is_checked() {
                        "Checked"
                    } else {
                        "Not_checked"
                    }
                );
            }
            println!("--------------");
        });
        grid.end();
        Panel {
            grid,
            label,
            cb1,
            cb2,
            btn,
        }
    } // end default fn;
} // end impl Panel;

fltk::widget_extends!(Panel, Grid, grid);

fn main() {
    let a = app::App::default();
    let mut win = window::Window::default().with_size(800, 600);
    let mut grid = Grid::default_fill();
    grid.show_grid(true);
    grid.set_layout(1, 2);
    let mut panel1 = Panel::default();
    let mut panel2 = Panel::default();
    grid.set_widget(&mut *panel1, 0, 0);
    grid.set_widget(&mut *panel2, 0, 1);
    grid.end();
    win.end();
    win.make_resizable(true);
    win.show();

    a.run().unwrap();
}


examples/form.rs
use fltk::{prelude::*, *};
use fltk_grid::Grid;

struct Form {
    grid: Grid,
    name: input::Input,
    age: input::IntInput,
    occupation: input::Input,
    btn: button::Button,
}

impl Form {
    pub fn default() -> Self {
        let mut grid = Grid::default_fill();
        grid.set_layout(10, 5); // construct a new grid
        let name = input::Input::default();
        let age = input::IntInput::default();
        let occupation = input::Input::default();
        let btn = button::Button::default().with_label("Submit");
        let mut g = Self {
            grid,
            name,
            age,
            occupation,
            btn,
        };
        g.fill();
        g
    }

    fn fill(&mut self) {
        let grid = &mut self.grid;
        grid.show_grid(false); // set to true to see cell outlines
        let mut title = frame::Frame::default().with_label("Employee Form");
        title.set_frame(enums::FrameType::FlatBox);
        title.set_color(enums::Color::Red);
        title.set_label_color(enums::Color::White);
        grid.set_widget(
            // set_widget widgets
            &mut title,
            0,
            1..4,
        );
        grid.set_widget(&mut frame::Frame::default().with_label("Name"), 2, 1);
        grid.set_widget(&mut self.name, 2, 3);
        grid.set_widget(&mut frame::Frame::default().with_label("Age"), 4, 1);
        grid.set_widget(&mut self.age, 4, 3);
        grid.set_widget(&mut frame::Frame::default().with_label("Occupation"), 6, 1);
        grid.set_widget(&mut self.occupation, 6, 3);
        grid.set_widget(&mut self.btn, 8, 2);
    }

    fn register_default_callback(&mut self) {
        self.btn.set_callback({
            let name = self.name.clone();
            let age = self.age.clone();
            let occupation = self.occupation.clone();
            move |_| {
                println!("Name: {}", name.value());
                println!("Age: {}", age.value());
                println!("Occupation: {}", occupation.value());
            }
        });
    }

    pub fn resize(&mut self, x: i32, y: i32, w: i32, h: i32) {
        self.grid.resize(x, y, w, h); // determine how it's resized
    }
}

fn main() {
    let a = app::App::default().with_scheme(app::Scheme::Gtk);
    let mut win = window::Window::default().with_size(500, 300);
    let mut form = Form::default();
    form.register_default_callback();
    win.end();
    win.make_resizable(true);
    win.show();

    win.resize_callback(move |_, _, _, w, h| form.resize(0, 0, w, h));

    a.run().unwrap();
}
